src/Compilers/Core/Portable/CommandLine/EditorConfig.cs

```diff
@@ -0,0 +1,127 @@
+                    {
+                        continue;
+                    }
+
+                    if (IsComment(line))
+                    {
+                        continue;
+                    }
+
+                    var sectionMatches = _sectionMatcher.Matches(line);
+                    if (sectionMatches.Count > 0 && sectionMatches[0].Groups.Count > 0)
+                    {
+                        var sectionName = sectionMatches[0].Groups[1].Value;
+                        if (!string.IsNullOrEmpty(sectionName))
+                        {
+                            activeSection = new Section(sectionName);
+                            editorConfig.NamedSections.Add(activeSection);
+                        }
+                        continue;
+                    }
+
+                    var propMatches = _propertyMatcher.Matches(line);
+                    if (propMatches.Count > 0 && propMatches[0].Groups.Count > 1)
+                    {
+                        var key = propMatches[0].Groups[1].Value;
+                        var value = propMatches[0].Groups[2].Value;
+                        if (!string.IsNullOrEmpty(key))
+                        {
+                            activeSection.Properties[key.Trim()] = value?.Trim();
+                        }
+                        continue;
+                    }
+                }
+            }
+
+            return editorConfig;
+        }
+
+        private static bool IsComment(string line)
+        {
+            for (int i = 0; i < line.Length; i++)
+            {
+                char c = line[i];
+                if (!Char.IsWhiteSpace(c))
+                {
+                    return c == '#' || c == ';';
+                }
+            }
+
+            return false;
+        }
+
+        internal class Section
+        {
+            public Section(string name)
+            {
+                this.Name = name;
+                // N.B. The editorconfig documentation is quite loose on property interpretation.
+                // Specifically, it says:
+                //      Currently all properties and values are case-insensitive.
+                //      They are lowercased when parsed.
+                // This is not a mandate for case-insensitive parsing, but a strong suggestion.
+                // As for *which* case-insensitive comparison, we use CaseInsensitiveComparison.Comparer,
+                // which is the recommended Unicode comparison for case-insensitive identifiers. The
+                // implementation performs the equivalent of a lower-case comparsion. The
+                // editorconfig file is defined as using a UTF-8 encoding, making Unicode comparison
+                // relevant.
+                this.Properties = new Dictionary<string, string>(CaseInsensitiveComparison.Comparer);
+            }
+
+            public string Name { get; }
+
+            public IDictionary<string, string> Properties { get; }
```

Comment:
@agocke Just an advisement about the conversation going on over here: https://github.com/dotnet/roslyn/issues/23391

Specifically, we've encountered some undefined behavior issues because some Roslyn features want to know the specific order of items in .editor_config.  This is because some features want to have a "first applicable editor_config entry 'wins'" strategy.  If we are to supply editorconfig support at the CodeAnalysis layer, it would be nice to be able to provide that.

I'll loop you into that issue.

Reply:
> This is because some features want to have a "first applicable editor_config entry 'wins'" strategy. 

Naively, that seems like the opposite of what users will expect. EditorConfig files are imported and parsed from root downward (to the current directory). The standard behavior according to the editorconfig "spec" is last definition wins.

> When opening a file, EditorConfig plugins look for a file named .editorconfig in the directory of the opened file and in every parent directory. A search for .editorconfig files will stop if the root filepath is reached or an EditorConfig file with root=true is found.

> EditorConfig files are read top to bottom and the closest EditorConfig files are read last. Properties from matching EditorConfig sections are applied in the order they were read, so properties in closer files take precedence.

Reply:
Let me rephrase:  the IDE wants to use order to define some behavioral contract.  Whether that is "first wins" or "last wins" and how that interacts with rules that are specified from outside-in, still needs to be committed to.  If you want, you can def provide feedback on the linked bug where this is being discussed.  The critical bit for IDE is that the API for obtaining this information be rich enough to at least supply whatever contract is decided upon, without depending on undefined behavior.